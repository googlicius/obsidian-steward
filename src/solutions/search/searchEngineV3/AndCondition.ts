import { Condition, ConditionResult } from './Condition';

/**
 * Composite condition for AND logic.
 */
export class AndCondition extends Condition {
  private conditions: Condition[];

  constructor(...conditions: Condition[]) {
    super();
    this.conditions = conditions;
  }

  async evaluate() {
    if (this.conditions.length === 0) return new Map();

    // Start with the first condition's results
    let result = await this.conditions[0].injectContext(this.context).evaluate();

    // Intersect with subsequent conditions, combining scores (e.g., sum for cumulative relevance)
    for (let i = 1; i < this.conditions.length; i++) {
      const nextMap = await this.conditions[i].injectContext(this.context).evaluate();
      const intersected = new Map<number, ConditionResult>();

      for (const [docId, resultData] of result) {
        const nextResult = nextMap.get(docId);
        if (nextResult !== undefined) {
          const combinedScore = resultData.score + nextResult.score; // Sum scores; adjust logic if needed (e.g., Math.max(score, nextScore) or score * nextScore for multiplicative)
          intersected.set(docId, {
            document: resultData.document,
            score: combinedScore,
          });
        }
      }
      result = intersected;
    }

    return result;
  }
}
