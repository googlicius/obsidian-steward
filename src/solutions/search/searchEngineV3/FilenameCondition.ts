import { Condition, ConditionResult } from './Condition';
import { ParsedRegexPattern } from '../types';
import { similarity } from 'src/utils/similarity';

const SIMILARITY_THRESHOLD = 0.7;

/**
 * Condition for filtering by filename.
 */
export class FilenameCondition extends Condition {
  constructor(private names: string[]) {
    super();
  }

  /**
   * Parse a regex pattern string and determine the search type and original name
   * Handles patterns like ^name$ (exact), ^name (startsWith), or name (contains)
   */
  private parseRegexPattern(pattern: string): ParsedRegexPattern {
    // Check if pattern is for exact match: ^name$
    if (pattern.startsWith('^') && pattern.endsWith('$')) {
      return {
        originalName: pattern.slice(1, -1),
        searchType: 'exact',
      };
    }

    // Check if pattern is for starts with: ^name
    if (pattern.startsWith('^')) {
      return {
        originalName: pattern.slice(1),
        searchType: 'startsWith',
      };
    }

    // Default case is contains: name
    return {
      originalName: pattern,
      searchType: 'contains',
    };
  }

  async evaluate() {
    const result = new Map<number, ConditionResult>();

    for (const name of this.names) {
      const terms = this.context.nameTokenizer.getUniqueTerms(name);
      if (terms.length === 0) continue;

      // Get document IDs from term entries
      const termEntries = await this.context.documentStore.getTermsByValue(terms);

      // Extract unique document IDs from term entries
      const documentIds = [...new Set(termEntries.map(entry => entry.documentId))];

      // If no matching documents found, continue to next name
      if (documentIds.length === 0) continue;

      // Fetch the documents for these IDs
      const documents = await this.context.documentStore.getDocumentsByIds(documentIds);

      // Filter documents by similarity score
      for (const doc of documents) {
        const parsedName = this.parseRegexPattern(name);
        let score = 0;

        switch (parsedName.searchType) {
          case 'contains':
            if (doc.fileName.toLowerCase().includes(parsedName.originalName.toLowerCase())) {
              score = 1.0;
            }
            break;

          case 'exact':
            score = similarity(parsedName.originalName, doc.fileName);
            break;

          case 'startsWith':
            if (doc.fileName.toLowerCase().startsWith(parsedName.originalName.toLowerCase())) {
              score = 1.0;
            }
            break;

          default:
            break;
        }

        // If similarity is above threshold, add to matched documents with score
        if (score >= SIMILARITY_THRESHOLD) {
          result.set(doc.id as number, {
            document: doc,
            score: score,
          });
        }
      }
    }

    return result;
  }
}
