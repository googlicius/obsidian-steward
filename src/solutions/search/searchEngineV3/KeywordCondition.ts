import { getQuotedQuery } from 'src/utils/getQuotedQuery';
import { Condition, ConditionResult } from './Condition';
import { ExactPhraseMatch } from '../types';
import { IndexedDocument, IndexedTerm, TermSource } from 'src/database/SearchDatabase';

const TERM_MATCH_THRESHOLD = 0.7;

/**
 * Condition for keyword search.
 */
export class KeywordCondition extends Condition<IndexedDocument> {
  constructor(private keywords: string[]) {
    super();
  }

  /**
   * Check if a keyword is an exact phrase (wrapped in quotes)
   * @param keyword The keyword to check
   * @returns The exact phrase details or null if not an exact phrase
   */
  private checkForExactPhrase(keyword: string): ExactPhraseMatch | null {
    const quotedContent = getQuotedQuery(keyword);

    if (quotedContent) {
      const phrase = quotedContent;
      const tokens = this.context.contentTokenizer.getUniqueTerms(phrase);
      return {
        originalPhrase: phrase,
        tokens,
      };
    }

    return null;
  }

  /**
   * Get term entries that match the specified content terms
   */
  private async getTermEntriesForContent(terms: string[]): Promise<IndexedTerm[]> {
    // Pre-compute arrays for better performance
    // const documentIdArray = scopedDocuments.map(doc => doc.id as number);
    // const folderIdArray = folders.map(folder => folder.id as number);

    // Apply filtering
    return (
      this.context.documentStore.terms
        .where('term')
        .anyOf(terms)
        .and(item => item.source === TermSource.Content)
        // .and(item => this.isDocumentMatch(item.documentId, documentIdArray, scopedDocuments.length))
        // .and(item => this.isFolderMatch(item.folderId, folderIdArray, folders.length))
        .toArray()
    );
  }

  /**
   * Check if a document has terms appearing consecutively
   * @param termsMap Map of terms to their positions in the document
   * @param tokens The tokens to check for consecutive positions
   * @returns True if the terms appear consecutively, false otherwise
   */
  private hasConsecutiveTerms(termsMap: Map<string, number[]>, tokens: string[]): boolean {
    if (tokens.length <= 1) {
      return true; // Single token is always "consecutive"
    }

    // Convert position arrays to Sets for O(1) lookup instead of O(n)
    const positionSets = new Map<string, Set<number>>();
    for (const token of tokens) {
      const positions = termsMap.get(token);
      if (!positions || positions.length === 0) {
        return false; // If any token is missing, no consecutive sequence possible
      }
      positionSets.set(token, new Set(positions));
    }

    // Get all positions for the first token
    const firstTokenPositions = positionSets.get(tokens[0]);
    if (!firstTokenPositions) {
      return false;
    }

    // For each position of the first token, check if it starts a consecutive sequence
    for (const startPos of firstTokenPositions) {
      let isConsecutive = true;

      // Check if all other tokens appear in consecutive positions
      for (let i = 1; i < tokens.length; i++) {
        const expectedPos = startPos + i;
        const tokenPositions = positionSets.get(tokens[i]);
        if (!tokenPositions) {
          isConsecutive = false;
          break;
        }

        // If the expected position is not found, this is not a consecutive sequence
        if (!tokenPositions.has(expectedPos)) {
          isConsecutive = false;
          break;
        }
      }

      // If we found a consecutive sequence, return true
      if (isConsecutive) {
        return true;
      }
    }

    // No consecutive sequence found
    return false;
  }

  /**
   * Handle exact phrase matching
   * @param exactPhrase The exact phrase to match
   * @param documentsMap Map to store matched documents
   */
  private async handleExactPhraseMatch(
    exactPhrase: ExactPhraseMatch,
    documentsMap: Map<number, ConditionResult<IndexedDocument>>
  ): Promise<void> {
    const { originalPhrase, tokens } = exactPhrase;

    if (tokens.length === 0) {
      return;
    }

    // Get term entries for all tokens in the phrase
    const termEntries = await this.getTermEntriesForContent(tokens);

    if (termEntries.length === 0) {
      return;
    }

    // Group by document ID
    const documentTermsMap = new Map<number, Map<string, number[]>>();

    // For each term entry, store the positions by document and term
    for (const entry of termEntries) {
      const { documentId, term, positions } = entry;

      if (!documentTermsMap.has(documentId)) {
        documentTermsMap.set(documentId, new Map());
      }

      const termsMap = documentTermsMap.get(documentId);
      if (!termsMap?.has(term)) {
        termsMap?.set(term, []);
      }

      termsMap?.get(term)?.push(...positions);
    }

    // Get documents that have all the tokens
    const potentialDocIds = Array.from(documentTermsMap.keys()).filter(docId => {
      const termsMap = documentTermsMap.get(docId);
      // Check if document has all tokens from the phrase
      return tokens.every(token => termsMap?.has(token));
    });

    if (potentialDocIds.length === 0) {
      return;
    }

    // For each potential document, check if it contains the exact phrase
    const exactMatchDocIds: number[] = [];

    for (const docId of potentialDocIds) {
      const termsMap = documentTermsMap.get(docId);
      if (!termsMap) continue;

      // Check for consecutive positions
      if (this.hasConsecutiveTerms(termsMap, tokens)) {
        exactMatchDocIds.push(docId);
      }
    }

    if (exactMatchDocIds.length === 0) {
      return;
    }

    // Fetch documents for the exact match IDs
    const exactMatchDocs = await this.context.documentStore.getDocumentsByIds(exactMatchDocIds);

    // Merge into overall result map
    for (const doc of exactMatchDocs) {
      const docId = doc.id as number;

      if (documentsMap.has(docId)) {
        // Document already exists, update score and keywords matched
        const existingResult = documentsMap.get(docId);
        if (existingResult) {
          existingResult.score += 10.0; // High score for exact matches
          existingResult.keywordsMatched = [
            ...(existingResult.keywordsMatched || []),
            originalPhrase,
          ];
        }
      } else {
        // New document, add to map
        documentsMap.set(docId, {
          document: doc,
          score: 10.0, // High score for exact matches
          keywordsMatched: [originalPhrase],
        });
      }
    }
  }

  /**
   * Group term entries by document
   */
  private groupTermEntriesByDocument(termEntries: IndexedTerm[]): Map<number, Set<string>> {
    const documentTermMap = new Map<number, Set<string>>();

    for (const entry of termEntries) {
      const { documentId, term } = entry;

      if (!documentTermMap.has(documentId)) {
        documentTermMap.set(documentId, new Set());
      }

      documentTermMap.get(documentId)?.add(term);
    }

    return documentTermMap;
  }

  /**
   * Get document IDs that meet the term match threshold
   */
  private getQualifiedDocumentIds(
    documentTermMap: Map<number, Set<string>>,
    terms: string[]
  ): number[] {
    const qualifiedIds: number[] = [];

    for (const [documentId, docTermsSet] of documentTermMap.entries()) {
      if (docTermsSet.size / terms.length >= TERM_MATCH_THRESHOLD) {
        qualifiedIds.push(documentId);
      }
    }

    return qualifiedIds;
  }

  async evaluate() {
    const documentsMap = new Map<number, ConditionResult<IndexedDocument>>();

    for (const keyword of this.keywords) {
      // Check if this is an exact phrase match
      const exactPhrase = this.checkForExactPhrase(keyword);

      if (exactPhrase) {
        // Handle exact phrase matching
        await this.handleExactPhraseMatch(exactPhrase, documentsMap);
        continue;
      }

      // Regular keyword matching (existing code)
      const terms = this.context.contentTokenizer.getUniqueTerms(keyword);
      if (terms.length === 0) {
        continue;
      }

      const termEntries = await this.getTermEntriesForContent(terms);
      const documentTermMap = this.groupTermEntriesByDocument(termEntries);

      // Get qualified document IDs for this keyword
      const keywordDocIds = this.getQualifiedDocumentIds(documentTermMap, terms);

      if (keywordDocIds.length === 0) {
        continue;
      }

      // Fetch documents for these IDs
      const keywordDocuments = await this.context.documentStore.getDocumentsByIds(keywordDocIds);

      // Calculate scores for this keyword using pre-tokenized terms
      const scoredDocumentsMap = await this.context.scoring.calculateDocumentScores(
        keywordDocuments,
        terms
      );

      // Filter out documents with 0 proximity bonus and merge into overall result map
      for (const [docId, scoredDoc] of scoredDocumentsMap.entries()) {
        // Skip documents with 0 proximity bonus
        if (scoredDoc.proximityBonus === 0) {
          continue;
        }

        if (documentsMap.has(docId)) {
          const existingResult = documentsMap.get(docId);
          if (existingResult) {
            existingResult.score += scoredDoc.score;
            existingResult.keywordsMatched = [...(existingResult.keywordsMatched || []), keyword];
          }
        } else {
          documentsMap.set(docId, {
            document: scoredDoc.document,
            score: scoredDoc.score,
            keywordsMatched: [keyword],
          });
        }
      }
    }

    return documentsMap;
  }
}
