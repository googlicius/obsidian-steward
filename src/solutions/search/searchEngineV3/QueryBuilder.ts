import { AndCondition } from './AndCondition';
import { Condition } from './Condition';
import { OrCondition } from './OrCondition';

/**
 * Generic QueryBuilder that allows adding any condition without business-specific methods.
 * Dependencies are provided through SearchContext to all conditions.
 */
export class QueryBuilder {
  private rootCondition: Condition;

  constructor() {
    this.rootCondition = new AndCondition();
  }

  /**
   * Add a condition with AND logic
   */
  and(condition: Condition): this {
    if (this.rootCondition instanceof AndCondition) {
      this.rootCondition = new AndCondition(...this.rootCondition['conditions'], condition);
    } else {
      this.rootCondition = new AndCondition(this.rootCondition, condition);
    }
    return this;
  }

  /**
   * Add a condition with OR logic
   */
  or(condition: Condition): this {
    if (this.rootCondition instanceof OrCondition) {
      this.rootCondition = new OrCondition(...this.rootCondition['conditions'], condition);
    } else {
      this.rootCondition = new OrCondition(this.rootCondition, condition);
    }
    return this;
  }

  build(): Condition {
    return this.rootCondition;
  }
}
