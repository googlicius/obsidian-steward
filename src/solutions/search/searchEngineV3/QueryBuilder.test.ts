import { AndCondition } from './AndCondition';
import { FilenameCondition } from './FilenameCondition';
import { KeywordCondition } from './KeywordCondition';
import { QueryBuilder } from './QueryBuilder';

describe('QueryBuilder', () => {
  it('should build an AND condition', () => {
    const queryBuilder = new QueryBuilder();
    queryBuilder.and(new FilenameCondition(['cat']));
    queryBuilder.and(new KeywordCondition(['black cat']));
    const condition = queryBuilder.build();
    expect(condition).toMatchSnapshot();
  });

  it('should build an OR condition', () => {
    // const orConditions: Condition[] = [];
    const queryBuilder = new QueryBuilder();

    queryBuilder.or(new FilenameCondition(['cat']));
    queryBuilder.or(new KeywordCondition(['black cat']));

    const condition = queryBuilder.build();

    expect(condition).toMatchSnapshot();
  });

  it('should build nested conditions', () => {
    const queryBuilder = new QueryBuilder();
    queryBuilder.or(
      new AndCondition(new FilenameCondition(['cat']), new KeywordCondition(['black cat']))
    );
    queryBuilder.or(
      new AndCondition(new FilenameCondition(['dog']), new KeywordCondition(['white dog']))
    );
    const condition = queryBuilder.build();
    expect(condition).toMatchSnapshot();
  });
});
