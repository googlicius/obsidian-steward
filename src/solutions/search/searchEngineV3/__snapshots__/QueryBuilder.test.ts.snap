// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder should build an AND condition 1`] = `
AndCondition {
  "conditions": [
    FilenameCondition {
      "names": [
        "cat",
      ],
    },
    KeywordCondition {
      "keywords": [
        "black cat",
      ],
    },
  ],
}
`;

exports[`QueryBuilder should build an OR condition 1`] = `
OrCondition {
  "conditions": [
    AndCondition {
      "conditions": [],
    },
    FilenameCondition {
      "names": [
        "cat",
      ],
    },
    KeywordCondition {
      "keywords": [
        "black cat",
      ],
    },
  ],
}
`;

exports[`QueryBuilder should build nested conditions 1`] = `
OrCondition {
  "conditions": [
    AndCondition {
      "conditions": [],
    },
    AndCondition {
      "conditions": [
        FilenameCondition {
          "names": [
            "cat",
          ],
        },
        KeywordCondition {
          "keywords": [
            "black cat",
          ],
        },
      ],
    },
    AndCondition {
      "conditions": [
        FilenameCondition {
          "names": [
            "dog",
          ],
        },
        KeywordCondition {
          "keywords": [
            "white dog",
          ],
        },
      ],
    },
  ],
}
`;
